bastion.cfn.yml

```diff
@@ -16,68 +16,290 @@ Parameters:
+                file = /var/log/secure
+                log_group_name = ${BastionSecureLogGroup}
+                log_stream_name = log
+                datetime_format = %b %d %H:%M:%S
+
+            "/etc/profile.d/init_google_authenticator.sh":
+              owner: root
+              group: root
+              content: !Sub |
+                #!/bin/bash -xe
+                if [ "${MFA}" == "true" ] && [ ! -e ~/.google_authenticator ]  &&  [ $USER != "root" ]; then
+                  echo -e "Initializing Google Authenticator\n"
+                  google-authenticator -t -d -f -r 3 -R 30 -w 3
+                  echo -e "Save the generated emergency scratch codes and use secret key or scan the QR code to register your device for multi-factor authentication.\n"
+                  echo -e "Login again using your ssh key pair and the generated one-time password on your registered device.\n"
+                  logout
+                fi
+
+            "/usr/local/sbin/configure_mfa.sh":
+              mode: "000550"
+              owner: root
+              group: root
+              content: !Sub |
+                #!/bin/bash -xe
+                if [ "${MFA}" == "true" ]; then
+                  echo "auth       required     pam_google_authenticator.so nullok" >> /etc/pam.d/sshd
+                  sed -e '/auth       substack     password-auth/ s/^#*/#/' -i /etc/pam.d/sshd
+                  sed -e '/ChallengeResponseAuthentication no/ s/^#*/#/' -i /etc/ssh/sshd_config
+                  sed -e '/#ChallengeResponseAuthentication yes/s/^#//' -i /etc/ssh/sshd_config
+                  echo >> /etc/ssh/sshd_config
+                  echo "AuthenticationMethods publickey,keyboard-interactive" >> /etc/ssh/sshd_config
+                  service sshd restart
+                fi
+                rm -f /usr/local/sbin/configure_mfa.sh
+
+          commands:
+            configure-mfa:
+              command: "/usr/local/sbin/configure_mfa.sh"
+
+          services:
+              sysvinit:
+                cfn-hup:
+                  enabled: true
+                  ensureRunning: true
+                  files:
+                    - /etc/cfn/cfn-hup.conf
+                    - /etc/cfn/hooks.d/cfn-auto-reloader.conf
+                awslogs:
+                  enabled: true
+                  ensureRunning: true
+                  files:
+                    - /etc/awslogs/awslogs.conf
+
     Properties:
       InstanceType: t2.micro
       KeyName: !Ref KeyName
+      NetworkInterfaces:
+        -
+          NetworkInterfaceId: !Ref BastionNetworkInterface
+          DeviceIndex: 0
+
       ImageId: !FindInMap [AMIMap, !Ref "AWS::Region", AMI]
-      SubnetId: !ImportValue
-        "Fn::Sub": "${NetworkStackName}-PublicSubnet1ID"
-      SecurityGroupIds:
-      - !ImportValue
-        "Fn::Sub": "${NetworkStackName}-BastionGroupID"
+      UserData:
+        Fn::Base64: !Sub |
+          #!/bin/bash -xe
+          yum update -y
```

Comment:
Probably out of scope for this pull, but reading this line made me wonder: how are patches automatically installed for long-running bastion hosts built from this template?

Reply:
MFA is phase 1, auto healing is phase 2, auto shutdown is phase 3 - the last phase is probably where this can probably be addressed.

Reply:
Note for later: I'd think about integrating EC2 Systems Manager for patch management.